projects/centre_console/src/race_switch.c

```diff
@@ -0,0 +1,146 @@
+#include "race_switch.h"
+
+#include <stdbool.h>
+
+#include "centre_console_events.h"
+#include "event_queue.h"
+#include "fsm.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "log.h"
+#include "status.h"
+#include "voltage_regulator.h"
+
+#define VOLTAGE_REGULATOR_ENABLE_ADDRESS \
+  { .port = GPIO_PORT_B, .pin = 6 }
+#define VOLTAGE_REGULATOR_MONITOR_ADDRESS \
+  { .port = GPIO_PORT_B, .pin = 7 }
+#define VOLTAGE_REGULATOR_DELAY_MS 25
+
+static GpioAddress s_race_switch_address = { .port = GPIO_PORT_A, .pin = 4 };
+
+// Must be redefined to pass as a pointer
+static GpioAddress s_voltage_regulator_enable_address = VOLTAGE_REGULATOR_ENABLE_ADDRESS;
+static GpioAddress s_voltage_regulator_monitor_address = VOLTAGE_REGULATOR_MONITOR_ADDRESS;
+
+static GpioSettings s_gpio_settings = { .state = GPIO_STATE_LOW,
+                                        .direction = GPIO_DIR_IN,
+                                        .resistor = GPIO_RES_NONE,
+                                        .alt_function = GPIO_ALTFN_NONE };
+static GpioSettings s_voltage_enable_gpio_settings = { .state = GPIO_STATE_LOW,
+                                                       .direction = GPIO_DIR_OUT,
+                                                       .resistor = GPIO_RES_NONE,
+                                                       .alt_function = GPIO_ALTFN_NONE };
+
+static InterruptSettings s_interrupt_settings = { .type = INTERRUPT_TYPE_INTERRUPT,
+                                                  .priority = INTERRUPT_PRIORITY_NORMAL };
+
+static void prv_voltage_monitor_error_callback(VoltageRegulatorError error, void *context) {
+  if (error == VOLTAGE_REGULATOR_ERROR_ON_WHEN_SHOULD_BE_OFF) {
+    LOG_WARN("Voltage Regulator Error: On when should be off");
+  } else {
+    LOG_WARN("Voltage Regulator Error: Off when should be on");
+  }
+}
+
+static VoltageRegulatorSettings s_voltage_regulator_settings = {
+  .enable_pin = VOLTAGE_REGULATOR_ENABLE_ADDRESS,
+  .monitor_pin = VOLTAGE_REGULATOR_MONITOR_ADDRESS,
+  .timer_callback_delay_ms = VOLTAGE_REGULATOR_DELAY_MS,
+  .error_callback = prv_voltage_monitor_error_callback,
+  .error_callback_context = NULL,
+};
+
+FSM_DECLARE_STATE(race_switch_off);
+FSM_DECLARE_STATE(race_switch_on);
+
+FSM_STATE_TRANSITION(race_switch_off) {
+  FSM_ADD_TRANSITION(RACE_STATE_ON, race_switch_on);
+}
+
+FSM_STATE_TRANSITION(race_switch_on) {
+  FSM_ADD_TRANSITION(RACE_STATE_OFF, race_switch_off);
```

Comment:
You should be transitioning on the events (i.e. `RACE_SWITCH_EVENT_ON` and `RACE_SWITCH_EVENT_OFF`), rather than the race states. Currently, this will transition if anyone raises an event with id 0 or 1 (currently the CAN RX and TX events!), which isn't what you want. Instead transition on the event IDs themselves and raise events with IDs `RACE_SWITCH_EVENT_ON` etc. You don't need the `s_event_lookup` lookup array at all I don't think.